### ISSUE TITLE
Ideas for Controller Design

### ISSUE THREAD

**Issue by fuelyourdev** [fuelyourdev/kvpr#1]
So here's my idea of what a UserController could look like, to kick off the conversation. I also built out or stubbed out a number of supporting classes.

```kotlin
@Controller("/users/:userId")
class UserInfoController(val userService: UserService) {

  @JsonSchema("""{
                .  "${"$"}schema": "http://json-schema.org/draft-04/schema#",
                .  "title": "Edit User Info",
                .  "description": "Edit User Info",
                .  "type": "object",
                .  "properties": {
                .    "header": {
                .      "allOf": [{ "${"$"}ref": "#/definitions/header" }],
                .      "properties": {
                .        "parameters": {
                .          "type": "object",
                .          "properties": {
                .            "userId":  {"type", "string"}
                .          },
                .          "required": ["userId"]
                .        }
                .      }
                .    }
                .  },
                .  "required": ["header"]
                .}""".trimMargin("."))
  fun get(userId: String?): Observable<Json> {
    if (userId != null) {
      return userService.get(userId).map{it.toJson()}
    } else {
      return userService.getAll().map{it.toJson()}
    }
  }

  @JsonSchema("""{
                .  "${"$"}schema": "http://json-schema.org/draft-04/schema#",
                .  "title": "Edit User Info",
                .  "description": "Edit User Info",
                .  "type": "object",
                .  "properties": {
                .    "header": {
                .      "allOf": [{ "${"$"}ref": "#/definitions/header" }],
                .      "properties": {
                .        "parameters": {
                .          "type": "object",
                .          "properties": {
                .            "userId":  {"type", "string"}
                .          },
                .          "required": ["userId"]
                .        }
                .      }
                .    },
                .    "body": {
                .      "type": "object",
                .      "properties": {
                .        "firstName": {"type": "string"},
                .        "lastName": {"type": "string"},
                .        "email": {"type": "string"}
                .      }
                .    }
                .  },
                .  "required": ["header", "body"]
                .}""".trimMargin(".")) 
  fun patch(userId: String, body: JsonObject): Observable<JsonObject> {
    return userService.edit(UserEdit.fromJson(userId, body)).map{it.toJson()}
  }
}

fun UserEdit.Companion.fromJson(userId: String, json: JsonObject): UserEdit {
  return UserEdit(userId,
                  json.getString("firstName"),
                  json.getString("lastName"),
                  json.getString("email"))
}

fun User.toJson(): JsonObject {
  return JsonObject("id" to this.userId,
                    "firstName" to this.firstName,
                    "lastName" to this.lastName,
                    "email" to this.email)
}

fun Array<User>.toJson(): JsonArray {
  return JsonArray(this.map {it.toJson()})
}

data class UserEdit(val userId: String,
                    val firstName: String?,
                    val lastName: String?,
                    val email: String?
) {
  companion object {}
}

data class User(val userId: String,
                val firstName: String,
                val lastName: String,
                val email: String
)

class JsonObject(vararg fields: Pair<String, Any>) : Json {
  val fields = mapOf(*fields)
  
  fun getString(key: String) : String? {
    return fields[key] as String?
  }
}

class JsonArray(vararg val entries: Any) : Json

interface Json

class UserService {
  
  fun get(userId: String) : Observable<User> {
    return User("", "", "", "")
  }
  
  fun getAll() : Observable<Array<User>> {
    return arrayOf()
  }
  
  fun edit(userEdit: UserEdit) : Observable<User> {
    return User("", "", "", "")
  }
}
```

**Comment by willisjtc** [comment]
Couple comments:
- i like the `@Url` annotation because I like the idea of being able to have multiple urls for a controller. e.g. the blog post example. Or let's consider another example of Users that when normal user views the users in the tenant it would probably be something like: `/tenants/:tenantId/users/:userId` but an admin for the site would possibly like to view all users and for one reason or another at `/users/:userId`
- I think I would want to keep the json schema in one file lie the mobile gateway project does. I think it would help keep the controller clean and focused. I could imagine a complicated json schema taking up 50 or more lines. Granted, I would be okay with having the json schema attached to a method in the controller file as long as we supported having one file of json schemas like in the mobile gateway

**Comment by fuelyourdev** [comment]
To point one, Colin discussed this a bit and and we think it would be nice to still have the clarity of a @Controller annotation, and we could simply make it support multiple values for a multi url situation, like so: `@Controller(urls = {"url1", "url2"})`. Let me know your thoughts on that.

To point two, I'll have examine the mobile gateway more in depth. Either that or if you could whip up a mock example, that could be helpful.

In my mind the controller is all about dealing with the specifics of the endpoint, url, and json, and as such business logic will be kept in the service/helper layer and kept out of the controller. 

So with that in mind, I can see the json schema being very relevant to the controller and nice to have in one place. On the flip side, json schema can get a little verbose, and could make things cluttered, and as such worthwhile to put in a separate file. So I can see benefits both ways, and would love to see a couple different proposals on how to handle it.

**Comment by willisjtc** [comment]
I was thinking of something like this for a controller:
`
@Controller(...)
class UsersController {
  fun get(val userId: String?, ...) {
    ...
  }
  
  fun post(val  userId: String?) {
    ...
  }

  ...
}
`

Something like this for the yaml file:
`
openapi: 3.0.0
servers:
  - url: 'http://web-gateway'
info:
  description: The web-gateway api docs
  version: 1.0.0
  title: web-gateway
  termsOfService: ''
  contact:
    email: koddle@koddle.com
  license:
    name: Apache 2.0
    url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
paths:
  /api/users/{userId}:
    get:
      summary: get a user
      operationId: getUsers
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/GetUsers'
        'XXX':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'
      parameters:
        - in: path
          name: contact_id
          description: The id of the user
          example: "1"
          schema:
            type: string
    post:
      summary: create a new user
      operationId: createUser
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserWithDetails'
        'XXX':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'
      parameters:
        - in: body
          name: body
          description: The user to be created
          required: true
          schema:
            $ref: '#/components/schemas/CreateUser'
`

Then in the framework we would do something like this:
`
class EndpointProcessor {
  openApiString = Load.fileByClassPath("mobile-gateway-openapi.yaml");
  SwaggerParseResult swaggerParseResult = new OpenAPIV3Parser().readContents(openApiString, null, null);
  openAPI = swaggerParseResult.getOpenAPI();
  schemas = openAPI.getComponents().getSchemas();
  updateFileSchema = schemas.get("UpdateFile");

   Schema togglesSchema = schemas.get("Toggles");
   toggles = togglesSchema.getProperties().keySet();

   schemaTester = new SchemaTester(openAPI);

  fun handleRequest(val requestContext: RequestContext) {
    JsonObject user = requestContext.body();
    JsonArray discrespancies = schemaTester.getDiscrepancies(createUserSchema, user, true);
     if (!discrepancies.isEmpty()) {
       handleDiscrepancies(requestContext, "contact", discrepancies);
       return;
     }
  }
}
`

**Comment by colinchilds** [comment]
I agree that the `@Controller` annotation should be able to take in a list of urls. I also dislike having the inline schema and would prefer either separate files like we do now in Canopy, or what Jonathan said. I'd be open to exploring the yaml approach.

**Comment by colinchilds** [comment]
I wonder if we should allow for custom marshalling functions on our Koddle framework so we can have each of the controller functions simply return an Observable and the marshaller will subscribe on it, add error handling with details about the method that failed, and serialize the emitted item to JSON
